spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteUtil.java

```diff
@@ -0,0 +1,247 @@
+    return new SparkWriteRequirements(distribution, ordering);
+  }
+
+  private static Distribution writeDistribution(Table table, DistributionMode mode) {
+    switch (mode) {
+      case NONE:
+        return Distributions.unspecified();
+
+      case HASH:
+        return Distributions.clustered(clustering(table));
+
+      case RANGE:
+        return Distributions.ordered(ordering(table));
+
+      default:
+        throw new IllegalArgumentException("Unsupported distribution mode: " + mode);
+    }
+  }
+
+  /** Builds requirements for copy-on-write DELETE, UPDATE, MERGE operations. */
+  public static SparkWriteRequirements copyOnWriteRequirements(
+      Table table, Command command, DistributionMode mode, boolean fanoutEnabled) {
+
+    if (command == DELETE || command == UPDATE) {
+      Distribution distribution = copyOnWriteDeleteUpdateDistribution(table, mode);
+      SortOrder[] ordering = writeOrdering(table, fanoutEnabled);
+      return new SparkWriteRequirements(distribution, ordering);
+    } else {
+      return writeRequirements(table, mode, fanoutEnabled);
+    }
+  }
+
+  private static Distribution copyOnWriteDeleteUpdateDistribution(
+      Table table, DistributionMode mode) {
+
+    switch (mode) {
+      case NONE:
+        return Distributions.unspecified();
+
+      case HASH:
+        if (table.spec().isPartitioned()) {
+          return Distributions.clustered(clustering(table));
+        } else {
+          return Distributions.clustered(FILE_CLUSTERING);
+        }
+
+      case RANGE:
+        if (table.spec().isPartitioned() || table.sortOrder().isSorted()) {
+          return Distributions.ordered(ordering(table));
+        } else {
+          return Distributions.ordered(EXISTING_FILE_ORDERING);
+        }
+
+      default:
+        throw new IllegalArgumentException("Unexpected distribution mode: " + mode);
+    }
+  }
+
+  /** Builds requirements for merge-on-read DELETE, UPDATE, MERGE operations. */
+  public static SparkWriteRequirements positionDeltaRequirements(
+      Table table, Command command, DistributionMode mode) {
+
+    if (command == UPDATE || command == MERGE) {
+      Distribution distribution = positionDeltaUpdateMergeDistribution(table, mode);
+      SortOrder[] ordering = concat(POSITION_DELETE_ORDERING, ordering(table));
+      return new SparkWriteRequirements(distribution, ordering);
+    } else {
+      Distribution distribution = positionDeltaDeleteDistribution(table, mode);
+      return new SparkWriteRequirements(distribution, POSITION_DELETE_ORDERING);
+    }
+  }
+
+  private static Distribution positionDeltaUpdateMergeDistribution(
+      Table table, DistributionMode mode) {
+
+    switch (mode) {
+      case NONE:
+        return Distributions.unspecified();
+
+      case HASH:
+        if (table.spec().isUnpartitioned()) {
+          Expression[] clustering = concat(SPEC_ID_PARTITION_FILE_CLUSTERING, clustering(table));
+          return Distributions.clustered(clustering);
```

Comment:
nit: can pull the return out of this branch statement, that way we wouldn't make two identical clustering variables in different scopes.

Reply:
I realize looking at future code you did this to avoid exceeding line length

Reply:
After I renamed the variable above like you suggested, it fits on one line now. I've inlined it.